### DISCUSSION TITLE
How to clean up duplicate commands?

### DISCUSSION THREAD

**Discussion by lexicalunit** [Rapptz/discord.py#7884] [Q&A]
For every one of my commands I see two options for it when I start typing. For example:
![Screen Shot 2022-04-13 at 6 01 39 PM](https://user-images.githubusercontent.com/1903876/163295943-48ac3f13-fc33-4185-ad87-f042357b1319.png)

One of these commands will work. The other one will cause an error:
```log
Ignoring exception in command tree:
Traceback (most recent call last):
  File ".../python3.9/site-packages/discord/app_commands/tree.py", line 756, in wrapper
    await self.call(interaction)
  File ".../python3.9/site-packages/discord/app_commands/tree.py", line 836, in call
    raise CommandNotFound(name, parents)
discord.app_commands.errors.CommandNotFound: Application command 'about' not found
```

Here's the code for this `/about` command:
```python
import logging

import discord
from discord.ext import commands
from discord import app_commands

from .. import SpellBot

logger = logging.getLogger(__name__)


class AboutCog(commands.Cog):
    def __init__(self, bot: SpellBot):
        self.bot = bot

    @app_commands.command(name="about", description="Get information about SpellBot.")
    async def about(self, interaction: discord.Interaction):
        ...

async def setup(bot: SpellBot):
    await bot.add_cog(AboutCog(bot), guild=bot.settings.DEBUG_GUILD_OBJECT)
```

Here's how I am syncing commands:
```python
if self.settings.DEBUG_GUILD:
    debug_guild = int(self.settings.DEBUG_GUILD)
    logger.info("syncing commands to debug guild: %s", debug_guild)
    await self.tree.sync(guild=discord.Object(id=debug_guild))
else:
    logger.info("syncing global commands")
    await self.tree.sync()
```

Note that I am not automatically syncing commands in my `setup_hook()` function as I've seen the FAQ against doing that many, many times.

I AM using a debug guild currently but in the past I have also synced globally (many weeks ago).

Note that previous to migrating to discord.py 2.0 I had been using discord-interactions v3.x with code like this:
```python
force_sync_commands: bool = ...
clean_commands: bool = ...
debug_guild: Optional[int] = None
if bot.settings.DEBUG_GUILD:
    debug_guild = int(bot.settings.DEBUG_GUILD)
    logger.info("using debug guild: %s", debug_guild)
bot.slash = SlashCommand(
    bot,
    debug_guild=debug_guild,
    sync_commands=force_sync_commands,
    delete_from_unused_guilds=clean_commands,
)
```

**Comment by Rapptz** [comment]
You're supposed to re-sync the guild with an empty set of commands. However, I forgot to write `tree.clear_commands` but I just added it in the latest commit. So you can now do `tree.clear_commands(guild=...)` followed by `await tree.sync(guild=...)`.

**Comment by lexicalunit** [comment]
**EDIT:** In this post I called `sync()` and `clear_commands()` out of order. Make sure to call `clear_commands()` first!

---

**Unedited original post:**

So I just ran:
```python
await self.tree.sync(guild=discord.Object(id=debug_guild))
self.tree.clear_commands(guild=discord.Object(id=debug_guild))
```

and then immediately afterwards I also ran:
```python
self.tree.clear_commands(guild=None)
await self.tree.sync()
```

But I still see one set of commands:
![Screen Shot 2022-04-14 at 9 16 23 AM](https://user-images.githubusercontent.com/1903876/163430996-b6ec40a2-9397-4104-906d-7d8c33bc0d1f.png)

When trying to run any of these lingering commands, I get this error:
```log
Ignoring exception in command tree:
Traceback (most recent call last):
  File "lib/python3.9/site-packages/discord/app_commands/tree.py", line 932, in wrapper
    await self.call(interaction)
  File "lib/python3.9/site-packages/discord/app_commands/tree.py", line 1071, in call
    command, options = self._get_app_command_options(data)
  File "lib/python3.9/site-packages/discord/app_commands/tree.py", line 962, in _get_app_command_options
    raise CommandNotFound(name, parents)
discord.app_commands.errors.CommandNotFound: Application command 'about' not found
```

**Reply by LostLuma** [reply]
You ran `clear_commands` and `sync` for the guild commands in the wrong order, meaning you synced them to Discord and then deleted them locally.

**Reply by lexicalunit** [reply]
> So you can now do `tree.clear_commands(guild=...)` followed by `await tree.sync(guild=...)`.

@LostLuma You're saying to run `sync` first _and then_ run `clear_commands` second? Because what I did was clear first, and then sync second. As shown in my copy/pasted code.

**Reply by LostLuma** [reply]
Your first code block has them in the wrong order.

**Reply by lexicalunit** [reply]
😅  I am very unobservant! Thank you.

**Comment by lexicalunit** [comment]
**EDIT:** This worked. But the suggestion by @Rapptz would also work. Just make sure to run `clear_commands()` first, and then `sync()` after that.

---
**Unedited original post:**

I think this is actually something weird lingering from my previous use of `discord-interactions` v3. I just ran:
```python
SlashCommand(
    bot,
    debug_guild=debug_guild,
    sync_commands=True,
    delete_from_unused_guilds=True,
)
```

And that seems to have cleaned up all the "broken" application commands. It was the inclusion of `delete_from_unused_guilds` that did the trick. That causes the following code to execute: https://github.com/interactions-py/library/blob/50066fc039b377f133c3008e8b7919d1fce2beb5/discord_slash/client.py#L546

```python
        if delete_from_unused_guilds:
            self.logger.info("Deleting unused guild commands...")
            other_guilds = [
                guild.id for guild in self._discord.guilds if guild.id not in cmds["guild"]
            ]
            # This is an extremly bad way to do this, because slash cmds can be in guilds the bot isn't in
            # But it's the only way until discord makes an endpoint to request all the guild with cmds registered.

            for guild in other_guilds:
                with suppress(discord.Forbidden):
                    existing = await self.req.get_all_commands(guild_id=guild)
                    if len(existing) != 0:
                        self.logger.debug(f"Deleting commands from {guild}")
                        await self.req.put_slash_commands(slash_commands=[], guild_id=guild)
```

🤷🏻‍♀️ 

Thanks for the help! Sorry to bother you with what I think is honestly a bug from another library.
